Public/Src/Utilities/Storage/InputChange/InputChangeList.cs

```diff
@@ -0,0 +1,195 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.ContractsLight;
+using System.IO;
+using System.Linq;
+using BuildXL.Storage.ChangeTracking;
+using BuildXL.Storage.Tracing;
+using BuildXL.Utilities.Instrumentation.Common;
+
+namespace BuildXL.Storage.InputChange
+{
+    /// <summary>
+    /// Class representing input change list.
+    /// </summary>
+    public sealed class InputChangeList : IObservable<ChangedPathInfo>
+    {
+        private readonly List<IObserver<ChangedPathInfo>> m_observers = new List<IObserver<ChangedPathInfo>>(2);
+        private readonly List<ChangedPathInfo> m_changedPaths = new List<ChangedPathInfo>();
+        private static char[] s_inputSeparator = new[] { '|' };
+
+        /// <summary>
+        /// Gets the list of input changes.
+        /// </summary>
+        public IEnumerable<ChangedPathInfo> ChangedPaths => m_changedPaths;
+
+        private InputChangeList()
+        {
+        }
+
+        /// <summary>
+        /// Creates and instance of <see cref="InputChangeList"/> from file.
+        /// </summary>
+        public static InputChangeList CreateFromFile(LoggingContext loggingContext, string path)
+        {
+            Contract.Requires(loggingContext != null);
+            Contract.Requires(!string.IsNullOrEmpty(path));
+
+            try
+            {
+                using (StreamReader reader = new StreamReader(path))
+                {
+                    return CreateFromStream(loggingContext, reader, path);
+                }
+            }
+            catch (IOException ioException)
+            {
+                Logger.Log.ExceptionOnCreatingInputChangeList(loggingContext, path, ioException.ToString());
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="InputChangeList"/> from a stream reader.
+        /// </summary>
+        public static InputChangeList CreateFromStream(LoggingContext loggingContext, TextReader reader, string optionalPath = null)
+        {
+            Contract.Requires(loggingContext != null);
+            Contract.Requires(reader != null);
+
+            InputChangeList inputChangeList = new InputChangeList();
+
+            try
+            {   
+                int lineNo = 0;
+
+                while (reader.Peek() >= 0)

```

Comment:
>e (reader.Peek() >= 0) [](start = 20, length = 22)

textreader will return null on reader.ReadLine when it is at the end, you don't have to peek....